### PR TITLE
Annotate DependencyInjection to make it linker friendly

### LINKED ISSUE #39745: Annotate DependencyInjection to make it linker friendly
The DependencyInjection library uses IL Emit and Reflection in order to create service implementations. However, this isn't linker friendly because the ILLinker will trim the constructors from these types, since it doesn't know anyone is using them.

We should annotate the DependencyInjection library with `DynamicallyAccessedMembers` attributes, so the ILLinker knows to preserve the constructors correctly. This should allow us to remove the `Microsoft.Extensions.*` assemblies from [Blazor's "type granularity" list](https://github.com/dotnet/aspnetcore/blob/fb28ce3ec2531327d6f98bfb7d44ad160393a925/src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Components.Wasm.targets#L372-L374), and instead fully trim these assemblies, just like the `System.*` assemblies.

I have a prototype of this [here](https://github.com/dotnet/runtime/compare/master...eerhardt:AnnotateDependencyInjection), where the size reduction is decent:

**Uncompressed IL file sizes:**
    - Before 2,425,856 bytes
    - After    2,360,320 bytes

**Compressed (.br) all file sizes in wwwroot\_framework:**
    - Before 1,464,786 bytes
    - After    1,443,696 bytes

We can see a ~20 KB size reduction in the Brotli compressed file sizes with this change.

cc @pranavkm @SteveSandersonMS @joperezr @layomia @davidfowl

### PR DESCRIPTION
**Pull request by eerhardt**
Fix #39745

With this change, the remaining (current) ILLinker warnings for the 2 DependencyInjection libraries are:

```
F:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection\src\ServiceLookup\Expressions\ExpressionResolverBuilder.cs(135,17): Trim analysis warning IL2006: Microsoft.Extensions.DependencyInjection.ServiceLookup.ExpressionResolverBuilder.VisitIEnumerable(IEnumerableCallSite,Object): Call to 'System.Reflection.MethodInfo.MakeGenericMethod' is not recognized.
F:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection\src\ServiceLookup\ILEmit\ILEmitResolverBuilder.cs(216,17): Trim analysis warning IL2006: Microsoft.Extensions.DependencyInjection.ServiceLookup.ILEmitResolverBuilder.VisitIEnumerable(IEnumerableCallSite,ILEmitResolverBuilderContext): Call to 'System.Reflection.MethodInfo.MakeGenericMethod' is not recognized.
F:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection\src\ServiceLookup\CallSiteFactory.cs(241,17): Trim analysis warning IL2006: Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateOpenGeneric(ServiceDescriptor,Type,CallSiteChain,Int32): Calling to 'System.Type.MakeGenericType' on unrecognized value.
```

### CODE REVIEW

File: src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs [hunk 1]
```diff
@@ -49,11 +52,9 @@ public static object CreateInstance(IServiceProvider provider, Type instanceType
 
             if (!instanceType.GetTypeInfo().IsAbstract)
             {
-                foreach (ConstructorInfo? constructor in instanceType
-                    .GetTypeInfo()
-                    .DeclaredConstructors)
+                foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
                 {
-                    if (!constructor.IsStatic && constructor.IsPublic)
```
Comment (davidfowl) [inline src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs:56]: Doesn't this change the runtime behavior?

Comment (eerhardt) [reply src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs:56]:
```C#
foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
```

That will only return `public` constructors.
What we were doing before was getting ALL constructors, but then filtering just the public constructors.
```C#
                foreach (ConstructorInfo? constructor in instanceType
                    .GetTypeInfo()
                    .DeclaredConstructors)
                {
                    if (!constructor.IsStatic && constructor.IsPublic)
                    {
                    }
                }
```

Comment (davidfowl) [reply src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs:56]: Got it, GetConstructors returns public constructors only.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/ServiceCollectionExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            return -1;
+        }
+
+        return 100;
+    }
+
+    private class ServiceA { }
+    private class ServiceB { }
+    private interface IServiceC { }
+    private class ServiceC : IServiceC { }
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/ServiceCollectionExtensionsTests.cs:30]: I would add one more complex test which is:

IA -> A
IB -> B(IA a)
IC -> C(IB b)

```C#
Resolve<IC>().Method();
```

**Review by davidfowl — APPROVED** [review]

**Review by maryamariyan — APPROVED** [review]

**Comment by davidfowl** [comment]
cc @dadhi @alexmg @seesharper @tillig @ENikS @ipjohnson @pakrym

More on this general feature here https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md

**Comment by ENikS** [comment]
Is this for DI libraries or client code? Will it remove client constructors marked for injection?

**Comment by davidfowl** [comment]
DI libraries. Today if you link the entire application it will remove constructors use in DI registrations because the linker can't see the which constructors are in use:

```C#
var services = new ServiceCollection();
services.AddSingleton<IFoo, Foo>();
var sp = services.BuildServiceProvider();
var foo = sp.GetService<IFoo>();
foo.Bar();
```

This would fail before this PR because Foo's constructor would be removed after linking. The annotations instruct the linker to preserve constructors of the generic second argument (the implementation type in this case).

**Comment by ENikS** [comment]
But what about constructors marked with [InjectionConstructor] attributes? I am assuming these will be gone as well?

**Comment by davidfowl** [comment]
@ENikS I don't know if there's a pattern to preserve those. Usually attribute only approaches require assembly scanning, which is also not linker friendly. It's possible that pattern is just incompatible and will result in the user needing to manually preserve those types.

cc @vitek-karas @MichalStrehovsky

**Comment by ENikS** [comment]
Without that MEF, MEF2, Unity and probable some other are not compatible with the pattern.

**Comment by eerhardt** [comment]
> Without that MEF, MEF2, Unity and probable some other are not compatible with the pattern.

Correct. We haven't gotten to making MEF and MEF2 linker friendly. It isn't scheduled for .NET 5, but it may be included in .NET 6.

**Comment by seesharper** [comment]
@davidfowl Thanks for letting us know about this.

So to preserve the constructors for a given type, the type needs to be mentioned in a method taking a `Type` parameter (or generic Type parameter )annotated with the `DynamicallyAccessedMembers`? If so, is that the only way to salvage these types? :)

**Comment by MichalStrehovsky** [comment]
> If so, is that the only way to salvage these types? :)

It's the only way that works with the dataflow analysis within the linker.

The problem is basically that code like:

```csharp
var  o = Activator.CreateInstance(someType);
```

may or may not work after trimming, depending on whether the default constructor for `someType` was preserved. By default, you'll get a warning at the time trimming runs for code like this. The warning will say "hey, there's a piece of reflection that I can't analyze. the code may or may not work at runtime".

You can get rid of that warning four ways:
1. You annotate things with the new attributes like in this pull request so that you'll end up with an annotation on `someType` that guarantees the default constructor was kept. This is the preferred solution, because it both makes the API "just work" (if linker can reason about the annotated pieces of code), and you got rid of the warning.
2. You suppress the warning with a suppression attribute and ensure the default constructor is kept through other means. E.g. add a direct call to the constructor somewhere. This is not great because by adding a warnings suppression, you're giving up on linker's warnings in the method and it's now your responsibility to make sure the API calls succeed after trimming.
3. You get rid of reflection. E.g. use a source generator.
4. You don't trim and mark the API with a new `RequiresUnreferencedCodeAttribute`. Linker will warn whenever a trimmed app calls into such API (this is to give a better user experience - instead of user getting a warning about a stray use of reflection in a library, they get a message that the library author made a deliberate choice not to be trimming friendly).

**Comment by davidfowl** [comment]
Does RequiresUnreferencedCodeAttribute disable trimming on the assembly/type/member specified?

**Comment by eerhardt** [comment]
> Does RequiresUnreferencedCodeAttribute disable trimming on the assembly/type/member specified?

No. It is a way to mark an API as unsafe for trimming. This lets the caller know that they should either call a different API or they need to do extra work to make this call safe (e.g. explicitly preserve types/methods/properties/etc).

For example `Assembly.Load(string | byte[])` methods will be marked with `RequiresUnreferencedCodeAttribute`.

**Comment by ENikS** [comment]
> You annotate things with the new attributes like in this pull request so that you'll end up with an annotation on someType that guarantees the default constructor was kept. This is the preferred solution, because it both makes the API "just work" (if linker can reason about the annotated pieces of code), and you got rid of the warning.

Does it have to be new attribute? Can you use existing MEF attributes?
These are available on every framework and would spare a lot of refactoring for project owners.

**Comment by davidfowl** [comment]
Nothing is being cut, those applications will still work with less aggressive trimming modes, they just won't be able to trim the assemblies completely (or at all).

**Comment by ENikS** [comment]
I understand, just trying to make sure this scenario is considered from the beginning and is not an afterthought when it is too late to change

**Comment by davidfowl** [comment]
See https://github.com/mono/linker/issues/1402

**Comment by ipjohnson** [comment]
@davidfowl thanks for the heads up. 

Can we apply `DynamicallyAccessedMembers` just to the api surface or do we need to put the attribute everywhere?

**Comment by davidfowl** [comment]
> Can we apply DynamicallyAccessedMembers just to the api surface or do we need to put the attribute everywhere?

You *can* apply it to just the API surface but the linker warns if it isn't applied transitively.

**Comment by seesharper** [comment]
Thanks @MichalStrehovsky for such a detailed explanation.👍

Is there any way we can start to test this? Is it possible to enable this using the preview 7 bits?

I've tried `dotnet publish -c Release -r osx-x64 --self-contained true -p:PublishSingleFile=false -p:PublishTrimmed=true` on a console app , but that did not seems to yield any warnings nor did it trim unreferenced types.

Finally I was just wondering if this is something that is mainly meant for Xamarin apps or do guys you see this as general feature that eventually can be applied to all .Net 5 apps?

**Comment by davidfowl** [comment]
@seesharper I made this for you https://github.com/davidfowl/LinkedAspNetCoreApplication. Run it with `dotnet publish -r osx-x64`

**Comment by seesharper** [comment]
Wow, that was fast. Thanks a lot, David. I'll have a look first thing tomorrow. Again, I gotta hand it to you guys. The way that you involve the community early on new features is just priceless. 👍💪 Really appreciate it

**Pull request merged**